src/Files.App/Services/Windows/WindowsRecycleBinService.cs

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2024 Files Community
+// Licensed under the MIT License. See the LICENSE.
+
+using System.Security.Principal;
+using Windows.Win32;
+
+namespace Files.App.Services
+{
+	/// <inheritdoc cref="IWindowsRecycleBinService"/>
+	public class WindowsRecycleBinService : IWindowsRecycleBinService
+	{
+		private readonly List<SystemIO.FileSystemWatcher> _watchers = [];
+
+		/// <inheritdoc/>
+		public event EventHandler<SystemIO.FileSystemEventArgs>? ItemAdded;
+
+		/// <inheritdoc/>
+		public event EventHandler<SystemIO.FileSystemEventArgs>? ItemDeleted;
+
+		/// <inheritdoc/>
+		public event EventHandler<SystemIO.FileSystemEventArgs>? ItemChanged;
+
+		/// <inheritdoc/>
+		public event EventHandler<SystemIO.FileSystemEventArgs>? ItemRenamed;
+
+		/// <inheritdoc/>
+		public event EventHandler<SystemIO.FileSystemEventArgs>? RefreshRequested;
+
+		/// <summary>
+  		/// Initializes an instance of <see cref="WindowsRecycleBinService"/> class.
+		/// </summary>
+		public WindowsRecycleBinService()
+		{
+			StartWatcher();
+		}
+
+		/// <inheritdoc/>
+  		public async Task<List<ShellFileItem>> GetAllRecycleBinFoldersAsync()
+		{
+			return (await Win32Helper.GetShellFolderAsync(Constants.UserEnvironmentPaths.RecycleBinPath, false, true, 0, int.MaxValue)).Enumerate;
+		}
+
+		/// <inheritdoc/>
+		public ulong GetSize()
+		{
+			return (ulong)Win32Helper.QueryRecycleBin().BinSize;
+		}
+
+		/// <inheritdoc/>
+		public bool IsRecycled(string? path)
+		{
+			return
+				!string.IsNullOrWhiteSpace(path) &&
+				RegexHelpers.RecycleBinPath().IsMatch(path);
+		}
+
+		/// <inheritdoc/>
+		public async Task<bool> IsRecyclableAsync(string? path)
+		{
+			if (string.IsNullOrEmpty(path) ||
+				path.StartsWith(@"\\?\", StringComparison.Ordinal))
+				return false;
+
+			var result = await FileOperationsHelpers.TestRecycleAsync(path.Split('|'));
+
+			return
+				result.Item1 &= result.Item2 is not null &&
+				result.Item2.Items.All(x => x.Succeeded);
+		}
+
+		/// <inheritdoc/>
+		public bool HasItems()
+		{
+			return Win32Helper.QueryRecycleBin().NumItems > 0;
+		}
+
+		/// <inheritdoc/>
+		public async Task<bool> DeleteAllAsync()
+		{
+			bool fRes = await Task.Run(() =>
+				PInvoke.SHEmptyRecycleBin(
+					new(),
+					string.Empty,
+					0x00000001 | 0x00000002 /*SHERB_NOCONFIRMATION | SHERB_NOPROGRESSUI*/)
+				.Succeeded);
+
+			return fRes;
+		}
+
+		/// <inheritdoc/>
+		public bool RestoreAll()
+		{
+			return true;
+		}
+
+		/// <inheritdoc/>
+		public void StartWatcher()
```

Comment:
@hishitetsu This service implementation is the very first time to use IWather. I’m feeling this is odd because StartWatcher and StopWatcher is public and you can call them via the interface.
Do you think I should create `RecycleBinWatcher : ITrashWatcher` and create a property of it in the service?
When you want to hook up the addition event of it yo have to access via `IWindowsRecycleBinService.Watcher.ItemAdded` for example.
This also can be said to IWindowsLibraryService and IStorageDevicesService and IMutableFolder (storage abstraction, you can ignore for now) in the future, such as `LibraryFolderWatcher/ShellFolderWatcher : IFolderWatcher` and `DeviceWatcher : IDeviceWatcher`.